selene-lib/src/rules/duplicate_keys.rs

```diff
@@ -0,0 +1,133 @@
+use crate::ast_util::range;
+
+use super::*;
+use std::{collections::HashMap, convert::Infallible};
+
+use full_moon::{
+    ast::{self, Ast},
+    tokenizer,
+    visitors::Visitor,
+};
+
+pub struct DuplicateKeysLint;
+
+impl Rule for DuplicateKeysLint {
+    type Config = ();
+    type Error = Infallible;
+
+    fn new(_: Self::Config) -> Result<Self, Self::Error> {
+        Ok(DuplicateKeysLint)
+    }
+
+    fn pass(&self, ast: &Ast, _: &Context) -> Vec<Diagnostic> {
+        let mut visitor = DuplicateKeysVisitor {
+            duplicates: Vec::new(),
+        };
+
+        visitor.visit_ast(&ast);
+
+        visitor
+            .duplicates
+            .iter()
+            .map(|duplicate| {
+                Diagnostic::new_complete(
+                    "duplicate_keys",
+                    format!("key `{}` is already declared", duplicate.name),
+                    Label::new(duplicate.position),
+                    Vec::new(),
+                    vec![Label::new_with_message(
+                        duplicate.original_declaration,
+                        format!("`{}` originally declared here", duplicate.name),
+                    )],
+                )
+            })
+            .collect()
+    }
+
+    fn severity(&self) -> Severity {
+        Severity::Warning
+    }
+
+    fn rule_type(&self) -> RuleType {
+        RuleType::Complexity
+    }
+}
+struct DuplicateKey {
+    name: String,
+    position: (usize, usize),
+    original_declaration: (usize, usize),
+}
+struct DuplicateKeysVisitor {
+    duplicates: Vec<DuplicateKey>,
+}
+
+/// Attempts to evaluate an expression key such as `"foobar"` in `["foobar"] = true` to a named identifier, `foobar`.
+/// Only works for string literal expression keys.
+fn key_expression_to_name(expression: &ast::Expression) -> Option<String> {
+    if let ast::Expression::Value { value, .. } = expression {
+        if let ast::Value::String(token) = &**value {
+            if let tokenizer::TokenType::StringLiteral { literal, .. } = token.token().token_type()
+            {
```

Comment:
Did Rustfmt do this? I wouldn't suspect it would move it to a new line.

Reply:
Yep it did